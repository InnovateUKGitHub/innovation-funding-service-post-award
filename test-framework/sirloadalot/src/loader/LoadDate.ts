export type LoadableDate =
  | RelativeDate
  | AbsoluteDate
  | (RelativeDate & AbsoluteDate);

export interface RelativeDate {
  /**
   * Set the date as the current date/time, with a specified offset.
   * Skipped inputs will retain the current value.
   */
  offset: Partial<DateInput> | "none";
}

export interface AbsoluteDate {
  /**
   * Set the date to a specific date/time.
   * Skipped inputs will retain the current value.
   */
  date: Partial<DateInput> | "now";
}

interface DateInput {
  /**
   * @asType integer
   */
  year: number;
  /**
   * @asType integer
   */
  month: number;
  /**
   * @asType integer
   */
  day: number;
  /**
   * @asType integer
   */
  hour: number;
  /**
   * @asType integer
   */
  minute: number;
}

interface LoadDateProps {
  date?: Date;
}

class LoadDate {
  date: Date;

  constructor({ date }: LoadDateProps) {
    // Clone the date, or create a new date.
    this.date = date ? new Date(date) : new Date();
  }

  setDate(input: LoadableDate): void {
    if ("date" in input) this.setAbsoluteDate(input);
    if ("offset" in input) this.setRelativeDate(input);
  }

  private setRelativeDate(input: RelativeDate): void {
    if (input.offset === "none") return;
    this.date.setUTCFullYear(
      this.date.getUTCFullYear() + (input.offset.year ?? 0),
      this.date.getUTCMonth() + (input.offset.month ?? 0),
      this.date.getUTCDate() + (input.offset.day ?? 0)
    );
    this.date.setUTCHours(
      this.date.getUTCHours() + (input.offset.hour ?? 0),
      this.date.getUTCMinutes() + (input.offset.minute ?? 0)
    );
  }

  private setAbsoluteDate(input: AbsoluteDate): void {
    if (input.date === "now") return;
    if (input.date.year !== undefined) {
      this.date.setUTCFullYear(input.date.year);
    }
    if (input.date.month !== undefined) {
      this.date.setUTCMonth(input.date.month - 1);
    }
    if (input.date.day !== undefined) {
      this.date.setUTCDate(input.date.day);
    }
    if (input.date.hour !== undefined) {
      this.date.setUTCHours(input.date.hour);
    }
    if (input.date.minute !== undefined) {
      this.date.setUTCMinutes(input.date.minute);
    }
  }

  toSalesforceDate(): string {
    return this.date.toISOString().replace("T", " ").substring(0, 16);
  }
}

export { LoadDate };
