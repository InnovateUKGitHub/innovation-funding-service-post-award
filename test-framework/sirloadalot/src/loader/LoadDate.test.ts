import { LoadDate, LoadableDate } from "./LoadDate";

describe("LoadDate", () => {
  const date = new Date("2020-04-23T12:34");

  test.each([
    ["offset by nothing", { offset: "none" } as LoadableDate],
    ["now", { date: "now" } as LoadableDate],
    ["a time an hour ago", { offset: { hour: -1 } } as LoadableDate],
    ["a day a month ago", { offset: { month: -1 } } as LoadableDate],
    ["a day 2 years ago", { offset: { year: -2 } } as LoadableDate],
    ["today midnight", { date: { hour: 0, minute: 0 } }],
    [
      "yesterday midnight",
      { date: { hour: 0, minute: 0 }, offset: { day: -1 } },
    ],
    [
      "christmas of the next year",
      {
        date: { month: 12, day: 25, hour: 0, minute: 0 },
        offset: { year: 1 },
      },
    ],
  ])("date should be %s", (_, loadableDate) => {
    const loader = new LoadDate({ date });
    loader.setDate(loadableDate as LoadableDate);
    expect(loader.toSalesforceDate()).toMatchSnapshot();
  });
});
