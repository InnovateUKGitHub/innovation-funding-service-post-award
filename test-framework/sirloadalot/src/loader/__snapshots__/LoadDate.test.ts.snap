// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadDate date should be a day 2 years ago 1`] = `"2018-04-23 11:34"`;

exports[`LoadDate date should be a day a month ago 1`] = `"2020-03-23 11:34"`;

exports[`LoadDate date should be a time an hour ago 1`] = `"2020-04-23 10:34"`;

exports[`LoadDate date should be christmas of the next year 1`] = `"2021-12-25 00:00"`;

exports[`LoadDate date should be now 1`] = `"2020-04-23 11:34"`;

exports[`LoadDate date should be offset by nothing 1`] = `"2020-04-23 11:34"`;

exports[`LoadDate date should be today midnight 1`] = `"2020-04-23 00:00"`;

exports[`LoadDate date should be yesterday midnight 1`] = `"2020-04-22 00:00"`;
